libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/datahandling/ReportBatchBuilder.java

```diff
@@ -2,56 +2,84 @@
+import org.mozilla.mozstumbler.service.utils.Zipper;
 
+import java.util.NoSuchElementException;
+import java.util.concurrent.ConcurrentLinkedQueue;
 
+/*
+ ReportBatchBuilder accepts MLS GeoSubmit JSON blobs and serializes them to
+ string form.  This is
+ */
 public class ReportBatchBuilder {
-
-    private static ILogger Log = (ILogger) ServiceLocator.getInstance().getService(ILogger.class);
-    private static String LOG_TAG = LoggerUtil.makeLogTag(ReportBatchBuilder.class);
-
-    // The max number of reports stored in the mCurrentReports. Each report is a GPS location plus wifi and cell scan.
-    // Once this size is reached, data is persisted to disk, mCurrentReports is cleared.
+    // The max number of reports stored in the mCachedReportBatches. Each report is a GPS location plus wifi and cell scan.
+    // Once this size is reached, data is persisted to disk, mCachedReportBatches is cleared.
     public static final int MAX_REPORTS_IN_MEMORY = 50;
-    public int wifiCount;
-    public int cellCount;
-    StringBuilder reportString;
-    int reportCount;
+
+    private ConcurrentLinkedQueue<MLSJSONObject> reportList = new ConcurrentLinkedQueue<MLSJSONObject>();
 
     public int reportsCount() {
-        return reportCount;
+        return reportList.size();
     }
 
-    String finalizeReports() {
-        final String kSuffix = "]}";
-        return reportString.toString() + kSuffix;
+    /*
+     This dequeues all collected reports and generates a stringified JSON blob
+
+     Setting preserveReports to true will ensure that the reports are not lost after being
+     converted to a string.  You almost certainly want to set that flag to false though as you'll
+     eat memory.
+     */
+    byte[] finalizeAndClearReports() {
+        return Zipper.zipData(generateJSON(false).getBytes());
     }
 
-    public void clearReports() {
-        reportCount = 0;
-        reportString = null;
+    private String generateJSON(boolean preserveReports) {
+
+        ConcurrentLinkedQueue<MLSJSONObject> jsonCollector = new ConcurrentLinkedQueue<MLSJSONObject>();
+        MLSJSONObject report = null;
+        final String kPrefix = "{\"items\":[";
+        StringBuilder reportString = new StringBuilder(kPrefix);
+        report = reportList.poll();
+        if (report != null) {
+            if (preserveReports) {
+                jsonCollector.add(report);
+            }
+            reportString.append(report.toString());
+        }
+
+        report = reportList.poll();
+        while (report != null) {
+            if (preserveReports) {
+                jsonCollector.add(report);
+            }
+            reportString.append("," + report.toString());
+            report = reportList.poll();
+        }
+
+        // Restore the ejected reports if necessary
+        if (preserveReports) {
+            report = jsonCollector.poll();
+            while (report != null) {
+                reportList.add(report);
```

Comment:
mReportList instead? I know, annoying. But I am so used to m* in our code base I code seriously confused reading this function understanding what is local and what is global.